### PR TITLE
SpiralGroundCurve: initial implementation.

### PR DESCRIPTION
**Pull request by agalbachicar**
# 🎉 New feature

Part of #265

## Summary

- Partial implementation of the SpiralGroundCurve.
- Test coverage for the entire class when the curve is built with normalized parameters.

Some TODOS in the implementation were left in favor of submitting a partial implementation that is properly covered with unit tests.  It is in favor of a more digestible and focused review of the math involved here.

## Test it

Via unit tests.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if it affects the public API)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/maliput_malidrive/road_curve/spiral_ground_curve.cc [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+namespace malidrive {
+namespace road_curve {
+namespace {
+// Number of coefficients for the series expansion of Fresnel sin and cosine.
+static constexpr size_t kNumCoefficients{10};
+
+// Computes the factorial of @p n at compile time whenever possible.
+// @param n A non-negative integer to compute the factorial to.
+// @return n!
+constexpr long Factorial(const long n) { return n <= 1L ? 1L : n * Factorial(n - 1L); }
```
Comment (stonier) [inline src/maliput_malidrive/road_curve/spiral_ground_curve.cc:47]: Should perhaps eventually land in `maliput::math`? Same with `PowerOf` and the `Fresnel` helpers?

**Review by stonier — COMMENTED** [review]
LGTM, but get Franco's eyeballs :)

**Review by stonier — APPROVED** [review]

File: src/maliput_malidrive/road_curve/spiral_ground_curve.h [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  /// @param heading0 The orientation of the tangent vector at @p xy0.
+  /// @param curvature0 Quantity which indicates the reciprocal of the
+  ///        turning radius of the arc at @p p0. A positive @p curvature0 makes a
+  ///        counterclockwise turn. It must be different from @p curvature1 by at least
+  ///        GroundCurve::kEpsilon. It must be of the same sign as @p curvature1.
+  /// @param curvature1 Quantity which indicates the reciprocal of the
+  ///        turning radius of the arc at @p p1. A positive @p curvature1 makes a
+  ///        counterclockwise turn. It must be different from @p curvature1 by at least
+  ///        GroundCurve::kEpsilon. It must be of the same sign as @p curvature0.
+  /// @param arc_length The spiral's length. It must be greate-r or equal to
```
Comment (francocipollone) [inline src/maliput_malidrive/road_curve/spiral_ground_curve.h:85]:
```suggestion
  /// @param arc_length The spiral's length. It must be greater or equal to
```

File: src/maliput_malidrive/road_curve/spiral_ground_curve.cc [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+  for (size_t i = 0; i < kNumCoefficients; ++i) {
+    x_terms[i] = PowerOf(x_terms[i], cos_exponents[i]);
+    y_terms[i] = PowerOf(y_terms[i], sin_exponents[i]);
+  }
+
+  return maliput::math::Vector2{
+      std::inner_product(x_terms.begin(), x_terms.end(), cos_coefficients.begin(), 0.),
+      std::inner_product(y_terms.begin(), y_terms.end(), sin_coefficients.begin(), 0.),
+  };
+}
```
Comment (francocipollone) [inline src/maliput_malidrive/road_curve/spiral_ground_curve.cc:132]: Maybe we could move all fresnel related stuff to a different header file (still private) so we can test it decoupled from the SpiralGroundCurve implementation.  (maybe not worth it to be added to maliput math yet.)

File: src/maliput_malidrive/road_curve/spiral_ground_curve.cc [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+  MALIDRIVE_THROW_UNLESS(arc_length_ >= GroundCurve::kEpsilon);
+  MALIDRIVE_THROW_UNLESS(p0_ >= 0.);
+  MALIDRIVE_THROW_UNLESS(p1_ - p0_ >= GroundCurve::kEpsilon);
+  MALIDRIVE_THROW_UNLESS(std::fabs(curvature1_ - curvature0_) >= GroundCurve::kEpsilon);
+  MALIDRIVE_THROW_UNLESS(std::signbit(curvature1_) == std::signbit(curvature0_));
+}
+
+double SpiralGroundCurve::DoPFromP(double xodr_p) const { return validate_p_(xodr_p); }
+
+double SpiralGroundCurve::NormalizedSpiralCoordinateAt(double p) const { return (p - p0_ + spiral_p0_) / norm_; }
```
Comment (francocipollone) [inline src/maliput_malidrive/road_curve/spiral_ground_curve.cc:170]: Reminder for me to ask you to walk me through the normalization/denormalization process here.

**Review by francocipollone — APPROVED** [review]
LGTM!

**Comment by agalbachicar** [comment]
Declining in favor of a new implementation.

**Pull request closed without merging**